python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,194 @@
+    from cudf._typing import Dtype
+    from cudf.core.column import ColumnBase
+    from cudf.core.frame import Frame
+
+
+class _Indexer:
+    # Indexer into a column (either a data column or index level).
+    #
+    # >>> df
+    #    a
+    # b
+    # 4  1
+    # 5  2
+    # 6  3
+    # >>> _Indexer("a", column=True).get(df)  # returns column "a" of df
+    # >>> _Indexer("b", index=True).get(df)  # returns index level "b" of df
+
+    def __init__(self, name: Any, column=False, index=False):
+        self.name = name
+        self.column, self.index = column, index
+
+    def get(self, obj: Frame) -> ColumnBase:
+        # get the column from `obj`
+        if self.column:
+            return obj._data[self.name]
+        else:
+            if obj._index is not None:
+                return obj._index._data[self.name]
+        raise KeyError()
+
+    def set(self, obj: Frame, value: ColumnBase):
+        # set the colum in `obj`
+        if self.column:
+            obj._data[self.name] = value
+        else:
+            if obj._index is not None:
+                obj._index._data[self.name] = value
+            else:
+                raise KeyError()
+
+
+def _frame_select_by_indexers(
+    frame: Frame, indexers: Iterable[_Indexer]
+) -> Frame:
+    # Select columns from the given `Frame` using `indexers`,
+    # and return a new `Frame`.
+    index_data = frame._data.__class__()
+    data = frame._data.__class__()
+
+    for idx in indexers:
+        if idx.index:
+            index_data[idx.name] = idx.get(frame)
+        else:
+            data[idx.name] = idx.get(frame)
+
+    result_index = cudf.Index._from_data(index_data) if index_data else None
+    result = cudf.core.frame.Frame(data=data, index=result_index)
+    return result
+
+
+def _match_join_keys(lcol: ColumnBase, rcol: ColumnBase, how: str) -> Dtype:
+    # cast the keys lcol and rcol to a common dtype
+
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # if either side is categorical, different logic
+    if isinstance(ltype, CategoricalDtype) or isinstance(
+        rtype, CategoricalDtype
+    ):
+        return _match_categorical_dtypes(ltype, rtype, how)
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return ltype
+
+    if (np.issubdtype(ltype, np.number)) and (np.issubdtype(rtype, np.number)):
+        common_type = (
+            max(ltype, rtype)
+            if ltype.kind == rtype.kind
+            else np.find_common_type([], (ltype, rtype))
+        )
+
+    elif np.issubdtype(ltype, np.datetime64) and np.issubdtype(
+        rtype, np.datetime64
+    ):
+        common_type = max(ltype, rtype)
+
+    if how == "left":
+        if rcol.fillna(0).can_cast_safely(ltype):
+            return ltype
+        else:
+            warnings.warn(
+                f"Can't safely cast column from {rtype} to {ltype}, "
+                "upcasting to {common_type}."
+            )
+
+    if common_type:
+        return common_type
+
+    return None
```

Comment:
I think making it here is cause to raise really. If the dtypes were the same, it should have returned `ltype` right?

Reply:
See comment above :-)